pysph/base/device_helper.py

```diff
@@ -308,40 +180,44 @@ def remove_prop(self, name):
     def resize(self, new_size):
         for prop in self.properties:
             stride = self._particle_array.stride.get(prop, 1)
-            self._data[prop].resize(new_size*stride)
-            setattr(self, prop, self._data[prop].array)
+            self._data[prop].resize(new_size * stride)
+            setattr(self, prop, self._data[prop])
 
     def align_particles(self):
-        tag_arr = self._data['tag'].array
+        tag_arr = self._data['tag']
 
         if len(tag_arr) == 0:
             self.num_real_particles = 0
             return
 
         num_particles = len(tag_arr)
 
-        new_indices = cl.array.empty(self._queue, num_particles,
-                                     dtype=np.int32)
-        num_real_particles = cl.array.empty(self._queue, 1,
-                                            dtype=np.int32)
-
-        align_particles_knl = GenericScanKernel(
-            self._ctx, np.int32,
-            arguments="__global int *tag_arr, __global int *new_indices, \
-                    __global int *num_real_particles, int num_particles",
-            input_expr="tag_arr[i] == 0",
-            scan_expr="a+b", neutral="0",
-            output_statement="""
-            int t = last_item + i - prev_item;
-            int idx = tag_arr[i] ? t : prev_item;
-            new_indices[idx] = i;
-            if(i == num_particles - 1)
-                num_real_particles[0] = last_item;
-            """
-        )
-
-        align_particles_knl(tag_arr, new_indices, num_real_particles,
-                            num_particles)
+        new_indices = array.empty(num_particles, dtype=np.int32,
+                                  backend=self.backend)
+        num_real_particles = array.empty(1, dtype=np.int32,
+                                         backend=self.backend)
+
+        @annotate(i='int', tag_arr='gintp', return_='int')
```

Comment:
This PR is nice but my only concern here is that each time `align_particles` is called (or any other of the methods using cpy), the function is created again, annotated and the code is generated.  All of the actual kernels should be cached carefully as these functions can be called quite often.

Reply:
Would be nice to have some sort of abstraction that handles this, as the same pattern is repeated many times.  I guess that since cpy doesn't handle methods, this is a bit harder than it should be but opening up wrapping methods leads to many potential issues like users accessing attributes of the class inside the methods.